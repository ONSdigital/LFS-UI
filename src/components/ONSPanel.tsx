import React, { Component } from 'react';

interface Props{
  label: string,
  children: any,
  header?: string,
  status? : string,
  spacious?: boolean,
  id?: string
}

export class ONSPanel extends Component <Props, {}>{

    render() {
      let className = "panel panel--" + (this.props.status === "success" ? "success" : this.props.status === "error" ? "error": "info") + " panel--simple " + (this.props.spacious ? "panel--spacious" : "");
      return (
        <div id={this.props.id} className={className}>
          {this.props.header !== undefined &&
            <div className="panel__header">
              <h1 className="panel__title u-fs-r--b">{this.props.header}</h1>
            </div>
          }
          <div className="panel__body">
              {this.props.children}
            </div>
        </div> 
      );
  }
}
