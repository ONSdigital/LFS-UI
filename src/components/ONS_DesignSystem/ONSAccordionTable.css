.nonSelectableTableRow:focus {
    -webkit-box-shadow: 0 0 0 3px #fd0;
    box-shadow: 0 0 0 3px #fd0;
    outline: none;
    padding: 0 0 0 1.85rem;
}

.selectableTableRow:focus {
    -webkit-box-shadow: 0 0 0 3px #fd0;
    box-shadow: 0 0 0 3px #fd0;
    outline: none;
    padding: 0 0 0 1.85rem;
}

.expandedRowEnabled .selectableTableRow:hover {
    cursor: pointer;
    -webkit-box-shadow: 0 0 0 3px #fe781f;
    box-shadow: 0 0 0 3px #fe781f;
    outline: none;
    padding: 0 0 0 1.85rem;
}

.expandedRow {
    padding: 10px 0 10px 0;
    outline: none;
    cursor: default;
    box-shadow: none;
}

.accordion-table-chevron {
    display: block;
    padding: 6px 0 6px 18px;
    margin: 0;
    width: 30px;
}

.accordion-table-chevron-svg {
    display: block;
    height: 15px;
    margin: 0;
}

.accordion-table-chevron-rotate {
    -webkit-transform:rotate(90deg);
    -moz-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    transform: rotate(90deg);
}