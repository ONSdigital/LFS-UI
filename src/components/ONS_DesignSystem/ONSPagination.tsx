import React, {Component} from 'react';

interface Props {
    steps: number,
    count: number,
    pageChange: (offset: number, steps: number) => void
}

interface State {
    page: number,
    maxPage: number
}

export class ONSPagination extends Component <Props, State> {
    constructor(props: Props) {
        super(props);
        let maxPage = Math.ceil(props.count / props.steps);
        this.state = {page: 1, maxPage: maxPage}
    }

    static getDerivedStateFromProps(nextProps: Props, prevState: State) {
        let maxPage = Math.ceil(nextProps.count / nextProps.steps);
        if (maxPage !== prevState.maxPage) {
            return {page: 1, maxPage: maxPage};
        } else return null;
    }


    createLinks = () => {

        let prev = [];
        let prevPages = [];
        for (let i = this.state.page - 1; i > 1 && i >= this.state.page - 2; i--) {
            prev.push(this.link(i));
            prevPages.push(i);
        }
        prev = prev.reverse();
        prevPages = prevPages.reverse();
        if (this.state.page !== 1 && this.state.page !== this.state.maxPage) {
            prev.push(this.link(this.state.page));
            prevPages.push(this.state.page)
        }
        let aft = [];
        let aftPages = [];
        for (let i = this.state.page + 1; i < this.state.maxPage && i <= this.state.page + 2; i++) {
            aft.push(this.link(i));
            aftPages.push(i);
        }
        let allPages = prevPages.concat(aftPages);
        let toReturn = prev.concat(aft);

        if (!allPages.includes(2)) {
            toReturn.unshift(<li key={"start_dots"} className="pagination__item pageination__item--gap">&hellip;</li>);
        }
        if (!allPages.includes(this.state.maxPage - 1)) {
            toReturn.push(<li key={"end_dots"} className="pagination__item pageination__item--gap">&hellip;</li>);
        }
        return toReturn
    };

    pageChange = (e: React.MouseEvent, page: number) => {
        if (page !== this.state.page) {
            this.props.pageChange((page - 1) * this.props.steps, this.props.steps);
            this.setState({page: page})
        }
    };

    previous = (e: React.MouseEvent) => {
        this.pageChange(e, this.state.page - 1)
    };

    next = (e: React.MouseEvent) => {
        this.pageChange(e, this.state.page + 1)
    };

    link = (page: number) => {
        return (<li key={page} style={{cursor: "pointer"}}
                    className={"pagination__item " + (this.state.page === page ? "pagination__item--current" : "")}><a
            onClick={(e) => this.pageChange(e, page)} className="pagination__link"
            aria-label={"Go to page " + page}>{page}</a></li>)
    };

    render() {
        return (
            <nav className="pagination "
                 arial-label={"Pagination (Page " + this.state.page + " of " + this.state.maxPage + ")"}>
                <div className="pagination__position" style={{marginBottom:'12px'}}>Page {this.state.page} of {this.state.maxPage}</div>
                <ul className="pagination__items" style={{marginLeft:'0px'}}>
                    {this.state.page > 1 &&
                    <li className="pagination__item pagination__item--previous" style={{backgroundColor:"#c1c9ca2e"}}>
                        <a style={{cursor: "pointer"}} onClick={(e) => this.previous(e)} className="pagination__link"
                           rel="prev" aria-label="Go to the previous page">Previous</a>
                    </li>
                    }
                    {this.link(1)}
                    {this.createLinks()}
                    {this.link(this.state.maxPage)}
                    {this.state.page < this.state.maxPage &&
                    <li className="pagination__item pagination__item--next" style={{backgroundColor:"#c1c9ca2e"}}>
                        <a style={{cursor: "pointer"}} onClick={(e) => this.next(e)} className="pagination__link"
                           rel="next" aria-label="Go to the next page">Next</a>
                    </li>
                    }
                </ul>
            </nav>
        );
    }
}
