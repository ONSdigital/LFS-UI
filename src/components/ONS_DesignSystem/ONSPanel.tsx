import React, {Component} from 'react';

interface Props{
  label: string,
  children: any,
  status? : string,
  spacious?: boolean,
  id?: string,
  hidden?: boolean
}

export class ONSPanel extends Component <Props, {}>{

    render() {
      let className = "panel panel--" + (this.props.status === "success" ? "success" : this.props.status === "error" ? "error": "info") + " panel--simple " + (this.props.spacious ? "panel--spacious" : "");
      return (
        <div id={this.props.id} className={className} hidden={this.props.hidden}>
          <div className="panel__body">
              {this.props.children}
          </div>
        </div> 
      );
  }
}
