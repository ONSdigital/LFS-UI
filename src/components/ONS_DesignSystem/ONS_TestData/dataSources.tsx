const tableData = ({
    "Rows": [{
        "source": "1",
        "week" : "1",
        "status": 1,
        "button": "",
        "button2": ""
    },
    {
        "source": "2",
        "week" : "2",
        "status": 1,
        "button": "",
        "button2": ""
    },
    {
        "source": "3",
        "week" : "3",
        "status": 1,
        "button": "2",
        "button2": ""
    },
    {
        "source": "4",
        "week" : "4",
        "status": 1,
        "button": "",
        "button2": ""
    },
    {
        "source": "5",
        "week" : "Monthly",
        "status": 1,
        "button": "",
        "button2": ""
    }],
        "Count":4
})

export {tableData}