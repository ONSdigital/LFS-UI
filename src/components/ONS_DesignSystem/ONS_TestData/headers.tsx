const DASHBOARD_HEADERS =
    [{
        label: "id",
        column_name: "id",
        filter: false,
        order: true
    }, {
        label: "status",
        column_name: "status",
        filter: false,
        order: true
    }];

export {DASHBOARD_HEADERS}