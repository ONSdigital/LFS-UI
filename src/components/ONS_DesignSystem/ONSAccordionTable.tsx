import React, {Component, Fragment} from 'react';
import {ONSPanel} from "./ONSPanel";
import update from 'immutability-helper';
import {ONSPagination} from "./ONSPagination";

interface Props {
    Headers: Header[]
    data: any[] | null
    id?: string
    Row: any
    expandedRowEnabled: boolean
    expandedRow?: any
    noDataMessage: string
    pagination?: boolean
    paginationSize?: number
    scrollable?: boolean
}

interface State {
    data: any[]
    slicedData: any[]
    offset: number
}

interface DashboardTableRow {
    id: number
    type: string
    period: string
    status: string
    year: number,
    rowExpanded: boolean
}

interface Header {
    label: string
    column_name: string
    filter: boolean
    order: boolean
}

export class ONSAccordionTable extends Component <Props, State> {

    constructor(props: Props) {
        super(props);

        this.state = {data: [], offset: 0, slicedData: []};
    }

    static getDerivedStateFromProps(nextProps: Props, prevState: State) {
        if (nextProps.data !== prevState.data && nextProps.data !== null) {
            return {data: nextProps.data, slicedData: nextProps.data.slice(0, nextProps.paginationSize)};
        } else return null;
    }

    pageChange = (offset: number, steps: number) => {
        this.setState({offset: offset});
        if (this.props.data === null) return;
        let slicedData: any[] = this.state.data.slice(offset, offset + steps);
        if (slicedData !== null) {
            this.setState({
                slicedData: slicedData
            })
        }
    };

    handleClickOnRow = (event: React.MouseEvent<HTMLTableRowElement, MouseEvent>, row: any, index: number) => {
        if (this.props.expandedRowEnabled) {
            row.rowExpanded = !row.rowExpanded;
            // @ts-ignore
            this.setState({data: update(this.state.data, {[index]: {$set: row}})})
        }
    };

    handleEnterKeyPressOnRow = (onClick: any, row: DashboardTableRow, index: number) => {
        if (onClick.key === 'Enter') {
            this.handleClickOnRow(onClick, row, index);
        }
    };

    Table = () => {
        return (
            <table id="basic-table" className="table ">
                <>
                    <thead className="table__head">
                    <tr className="table__row">
                        {
                            this.props.Headers.map((header: any, index: number) =>
                                <th key={index} scope="col" className="table__header ">{header.label}</th>
                            )
                        }
                    </tr>
                    </thead>
                    <tbody className="table__body">
                    {
                        this.state.data !== null && this.state.data.length !== 0 ?
                            this.state.slicedData.map((row: DashboardTableRow, index: number) =>
                                <Fragment key={(this.props.expandedRowEnabled ? row.id : index)}>
                                    <tr className={("table__row selectableTableRow " + (this.props.expandedRowEnabled ? "itemCursorHover " : ""))}
                                        onClick={((e) => this.handleClickOnRow(e, row, index))}
                                        tabIndex={0}
                                        onKeyPress={((e => this.handleEnterKeyPressOnRow(e, row, index)))}>
                                        <this.props.Row row={row}/>
                                    </tr>
                                    {
                                        this.props.expandedRowEnabled ?
                                            <tr hidden={!row.rowExpanded}>
                                                <td colSpan={this.props.Headers.length}
                                                    className="table__cell ">
                                                    <this.props.expandedRow row={row}/>
                                                </td>
                                            </tr>
                                            :
                                            <>
                                            </>
                                    }

                                </Fragment>
                            )
                            :
                            <tr>
                                <td colSpan={this.props.Headers.length} className="table__cell ">
                                    <ONSPanel label={'No Batches Matching the Criteria'}>
                                        <p>{this.props.noDataMessage}</p>
                                    </ONSPanel>
                                </td>
                            </tr>
                    }
                    </tbody>
                </>
            </table>
        )
    };

    render() {
        return (
            <>
                {
                    this.props.scrollable ?
                        <>
                            <ONSPanel label={" "}>
                                {"Scrollable table. There are " + this.props.Headers.length + " columns in this table. Some of the table may be off screen. Scroll or drag horizontally to bring into view. "}
                            </ONSPanel>
                            <div className="table-scrollable table-scrollable--on">
                                <div className="table-scrollable__content" tabIndex={0} role="region"
                                     aria-label={"Scrollable table. There are " + this.props.Headers.length + " columns in this table. Some of the table may be off screen. Scroll or drag horizontally to bring into view. "}>
                                    <br/>

                                    <this.Table/>
                                </div>
                            </div>
                        </>
                        :
                        <>
                            <this.Table/>
                        </>
                }
                {this.props.pagination ?
                    <ONSPagination
                        steps={(this.props.paginationSize !== undefined ? this.props.paginationSize : 20)}
                        count={this.state.data.length} pageChange={this.pageChange}/>
                    :
                    <br/>
                }
            </>
        );
    }
}
