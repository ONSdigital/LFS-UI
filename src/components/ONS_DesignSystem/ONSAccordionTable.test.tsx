import React from "react";
import Enzyme, {mount, shallow} from "enzyme";
import {ONSAccordionTable} from "./ONSAccordionTable";
import Adapter from "enzyme-adapter-react-16";
import {DASHBOARD_HEADERS} from "./ONS_TestData/headers";
import {ONSPagination} from "./ONSPagination";
import {ONSPanel} from "./ONSPanel";
import {fireEvent, render, screen} from "@testing-library/react";
import {default as flushPromises} from "../../tests/util/flushPromises";

interface tableRow {
    id: string
    status: string
}

describe("ONS Accordion Table Test", () => {
    Enzyme.configure({adapter: new Adapter()});

    let emptyData: any = [];

    const singleRowData = [{id: "1", status: "3"}];

    const manyRowData = [
        {id: 1, status: "1"},
        {id: 2, status: "2"},
        {id: 3, status: "3"}];

    const tableRow = (rowData: any) => {
        let row: tableRow = rowData.row;
        return (
            <>
                <td className="table__cell ">
                    {row.id}
                </td>
                <td className="table__cell ">
                    {row.status}
                </td>
            </>
        );
    };

    const DashboardExpandedRow = (rowData: any) => {
        let row: tableRow = rowData.row;
        return (
            <>
                <p>heyyyyyyy</p>
            </>
        );
    };

    const singleRowProps = {
        Headers: DASHBOARD_HEADERS,
        data: singleRowData,
        Row: tableRow,
        expandedRowEnabled: true,
        expandedRow: DashboardExpandedRow,
        noDataMessage: "data mate 1"
    };

    const falseExSingleRowProps = {
        Headers: DASHBOARD_HEADERS,
        data: singleRowData,
        Row: tableRow,
        expandedRowEnabled: false,
        expandedRow: DashboardExpandedRow,
        noDataMessage: "info in your data mate 2"
    };

    const emptyProps = {
        Headers: DASHBOARD_HEADERS,
        data: emptyData,
        Row: tableRow,
        expandedRowEnabled: false,
        expandedRow: DashboardExpandedRow,
        noDataMessage: "error in your data mate 3"
    };

    const trueEmptyProps = {
        Headers: DASHBOARD_HEADERS,
        data: emptyData,
        Row: tableRow,
        expandedRowEnabled: true,
        expandedRow: DashboardExpandedRow,
        noDataMessage: "no data mate 3"
    };

    const manyRowProps = {
        Headers: DASHBOARD_HEADERS,
        data: manyRowData,
        Row: tableRow,
        expandedRowEnabled: true,
        expandedRow: DashboardExpandedRow,
        noDataMessage: "no data mate 4",
        pagination: true,
        paginationSize: 1,
        scrollable: true
    };

    const manyRowPropsNoExpansion = {
        Headers: DASHBOARD_HEADERS,
        data: manyRowData,
        Row: tableRow,
        expandedRowEnabled: false,
        noDataMessage: "no data mate 4",
        pagination: false
    };

    const manyRowPropsNoExpansionPagination = {
        Headers: DASHBOARD_HEADERS,
        data: manyRowData,
        Row: tableRow,
        expandedRowEnabled: false,
        noDataMessage: "no data mate 4",
        pagination: true,
        paginationSize: 2,
    };

    const uManyRowProps = {
        //pagination is undefined
        Headers: DASHBOARD_HEADERS,
        data: manyRowData,
        Row: tableRow,
        expandedRowEnabled: true,
        expandedRow: DashboardExpandedRow,
        expandedAdditionalRows: true,
        noDataMessage: "no data mate 5",
        pagination: true,
        paginationSize: undefined
    };

    const captionProps = {
        Headers: DASHBOARD_HEADERS,
        data: singleRowData,
        Row: tableRow,
        expandedRowEnabled: true,
        expandedRow: DashboardExpandedRow,
        noDataMessage: "no data mate 1",
        caption: "Table Title"
    };

    function wrapper(render: any, props: any) {
        return render(
            <ONSAccordionTable
                {...props}>
            </ONSAccordionTable>
        );
    }

    it("should render correctly", () => expect(wrapper(shallow, singleRowProps).exists()).toEqual(true));

    it("should render with the correct headers", () => {
        expect(wrapper(mount, falseExSingleRowProps).find("ONSAccordionTable").getElement().props.Headers).toEqual(falseExSingleRowProps.Headers);
    });

    it("should render with the correct headers", () => {
        expect(wrapper(mount, trueEmptyProps).find("ONSAccordionTable").getElement().props.noDataMessage).toEqual(trueEmptyProps.noDataMessage);
    });

    it("should render with Caption above Table", () => {
        expect(wrapper(mount, captionProps).find("caption").text()).toEqual(captionProps.caption);
    });

    it("simulates click events", () => {
        let exWrapper = wrapper(mount, singleRowProps);
        let exRow = exWrapper.find("tr.selectableTableRow");
        exRow.simulate("click");
        expect(exRow.find("rowExpanded")).toBeTruthy();

        let unexWrapper = wrapper(mount, falseExSingleRowProps);
        let unexRow = unexWrapper.find("tr.nonSelectableTableRow");
        unexRow.simulate("click");
        expect(unexRow.find("rowExpanded")).not.toEqual(true);

    });

    it("simulates keypress events", () => {
        let thisWrapper = wrapper(mount, uManyRowProps);
        let thisRow = thisWrapper.find("tr.selectableTableRow").at(0);
        thisRow.simulate("keypress", {key: "Enter"});
        expect(thisRow.find("rowExpanded")).toBeTruthy();
        thisRow.simulate("keypress");
        expect(thisRow.find("rowExpanded")).not.toEqual(true);
    });

    it("simulates pageChange event", () => {
        let thisWrapper = wrapper(shallow, manyRowProps);
        let offset = thisWrapper.state("offset");
        thisWrapper.find(ONSPagination).dive().find("li.pagination__item--next").find("button").simulate("click");
        expect(thisWrapper.state("offset")).not.toEqual(offset);
    });

    it("should display a message if the data is empty", () => {
        const thisWrapper = wrapper(mount, emptyProps);
        expect(thisWrapper.find(ONSPanel).find("p").text()).toEqual(emptyProps.noDataMessage);
    });

    it("should reorder the table when you click a sortable column header ", async () => {
        wrapper(render, manyRowPropsNoExpansion);

        // ---------------------- Check default order ----------------------
        // Get list Items
        let list = screen.queryAllByTestId(/table-row/i);
        let listItemOne = list[0];
        let listItemThree = list[2];

        // Check if status column for first item is in default order
        let secondColumnValue = listItemOne.children[0].textContent;
        expect(secondColumnValue).toEqual("1");

        // Check if status column for third item has in default order
        secondColumnValue = listItemThree.children[0].textContent;
        expect(secondColumnValue).toEqual("3");

        // ---------------------- Change order to descending by second column ----------------------

        // Click on the header to change order
        await fireEvent.click(screen.getByText(/Status/i));
        await flushPromises();

        // Get list Items again
        list = screen.queryAllByTestId(/table-row/i);
        listItemOne = list[0];
        listItemThree = list[2];

        // Check if status column for first item is descending by Status column
        secondColumnValue = listItemOne.children[1].textContent;
        expect(secondColumnValue).toEqual("3");

        // Check if status column for third item is descending by Status column
        secondColumnValue = listItemThree.children[1].textContent;
        expect(secondColumnValue).toEqual("1");

        // ---------------------- Change order to ascending by second column ----------------------

        // Click on the header to change order Again
        await fireEvent.click(screen.getByText(/Status/i));
        await flushPromises();

        // Get list Items
        list = screen.queryAllByTestId(/table-row/i);
        listItemOne = list[0];
        listItemThree = list[2];

        // Check if status column for first item is ascending by Status column
        secondColumnValue = listItemOne.children[1].textContent;
        expect(secondColumnValue).toEqual("1");

        // Check if status column for third item is ascending by Status column
        secondColumnValue = listItemThree.children[1].textContent;
        expect(secondColumnValue).toEqual("3");
    });

    it("should reorder the table when you click a sortable column header across pages in a pagination table", async () => {
        wrapper(render, manyRowPropsNoExpansionPagination);

        // Click the next button in the pagination
        await fireEvent.click(screen.getByText(/Next/i));
        await flushPromises();

        // Get list Items for the next page
        let list = screen.queryAllByTestId(/table-row/i);
        let listItemOne = list[0];

        // Check if status column for first item on the second page is the default third item
        let secondColumnValue = listItemOne.children[1].textContent;
        expect(secondColumnValue).toEqual("3");

        // Click on the header to change order
        await fireEvent.click(screen.getByText(/Status/i));
        await flushPromises();


        // Get list Items for the next page
        list = screen.queryAllByTestId(/table-row/i);
        listItemOne = list[0];

        // list.forEach((item) => console.log(item.children[0].textContent + " - " + item.children[1].textContent))

        // Check if status column for first item on the second page is ascending by Status column
        secondColumnValue = listItemOne.children[1].textContent;
        expect(secondColumnValue).toEqual("1");
    });
});