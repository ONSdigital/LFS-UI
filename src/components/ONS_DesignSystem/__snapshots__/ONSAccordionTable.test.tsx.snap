// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ONS Accordion Table Test matches Snapshot 1`] = `ShallowWrapper {}`;
